### PR TITLE
[WIP] dataloader templates and tutorial for custom image loader

### PR DESCRIPTION
**Pull request by pengyuan-zhou**
- [ ] template revise

- [ ] tutorial code debug

- [ ] tutorial explanation to add

### CODE REVIEW

File: tutorial/Part 2- Custom image dataset loader.md [hunk 1]
```diff
@@ -0,0 +1,242 @@
...
+          readSample(features, imgTensors)
+          
+        DATASIZE = features[0].size
+        
+        val trainingData = Batch(
+            features.flatten().toFloatArray(),
+            longArrayOf(features.size.toLong(), DATASIZE!!.toLong())
+        )
+        val trainingImage = Batch(
+            imgTensors,// TODO: 25/08/2020
```
Comment (pengyuan-zhou) [inline tutorial/Part 2- Custom image dataset loader.md:70]: TODO:  convert tensor to float array

File: syftlib/src/main/java/org/openmined/syft/datasource/Dataset.kt [hunk 2]
```diff
@@ -0,0 +1,26 @@
+package org.openmined.syft.datasource
+
+@ExperimentalUnsignedTypes
+open class Dataset {
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/Dataset.kt:4]: This should be an interface.

File: syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+ * @param replacement (Boolean): samples are drawn on-demand with replacement if ``True``, default=``False``
+ * @param numSamples (Int): number of samples to draw, default=`len(dataset)`. This argument
+ *                          is supposed to be specified only when `replacement` is ``True``.
+ */
+class RandomSampler(override var dataSource: Dataset, var replacement: Boolean = false,
+                    var numSamples: Int = dataSource.len()) : Sampler(dataSource) {
+
+     fun numsamples() : Int{
+
+         if (numSamples == null) {
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt:19]: Use `numSample?.let {} :? `

File: syftlib/src/main/java/org/openmined/syft/datasource/Sampler.kt [hunk 4]
```diff
@@ -0,0 +1,12 @@
+package org.openmined.syft.datasource
+
+/**
+ * Base class for all Samplers.
+ * Every Sampler subclass has to provide an :meth:`__iter__` method, providing a
+ * way to iterate over indices of dataset elements, and a :meth:`__len__` method
+ * that returns the length of the returned iterators.
+ */
+open class Sampler (open var dataSource: Dataset){
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/Sampler.kt:9]: Interface

Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/Sampler.kt:9]: Alternatively, build using a `Template` pattern.

File: syftlib/src/main/java/org/openmined/syft/datasource/DataLoader.kt [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+ *                  at every epoch (default: ``False``).
+ * @param sampler (Sampler or Iterable, optional): defines the strategy to draw
+ *                  samples from the dataset. Can be any ``Iterable`` with ``__len__``
+ *                  implemented. If specified, :attr:`shuffle` must not be specified.
+ * @param dropLast (Boolean, optional): set to ``True`` to drop the last incomplete batch,
+ *                  if the dataset size is not divisible by the batch size. If ``False`` and
+ *                  the size of dataset is not divisible by the batch size, then the last batch
+ *                  will be smaller. (default: ``False``)
+ */
+class DataLoader(var dataset: Dataset, var batchsize: Int = 1, var shuffle: Boolean = false,
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/DataLoader.kt:24]: Either pass the `sampler` instance or `shuffle` but not both (I'd prefer the instance).

**Review by mccorby — CHANGES_REQUESTED** [review]

File: syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt (hunk 3 above)
Comment (pengyuan-zhou) [reply syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt:19]: like `numSamples?.let {return dataSource.len()}`  ?

Comment (pengyuan-zhou) [reply syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt:19]: > Use `numSample?.let {} :? `

like `numSamples?.let {return dataSource.len()}` ?

Comment (mccorby) [reply syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt:19]: It's more Kotlin idiomatic. Also the last condition should never happen (`return 0`).
```
fun getNumSamples() : Int{
	return numSamples?.let {
        numSamples
    } ?: 
        dataSource.size
}
```

File: syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * Samples elements randomly. If without replacement, then sample from a shuffled dataset.
+ * If with replacement, then user can specify :attr:`num_samples` to draw.
+ * @param dataSource (Dataset): dataset to sample from
+ * @param replacement (Boolean): samples are drawn on-demand with replacement if ``True``, default=``False``
+ * @param numSamples (Int): number of samples to draw, default=`len(dataset)`. This argument
+ *                          is supposed to be specified only when `replacement` is ``True``.
+ */
+class RandomSampler(override var dataSource: Dataset, var replacement: Boolean = false,
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt:14]: These `var` should be `val` as they are immutable.

Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt:14]: What is the role of `replacement`? It can be achieved with the nullity of `numSamples`

File: syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+                    var numSamples: Int = dataSource.len()) : Sampler(dataSource) {
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt:15]: `numSamples` should be declared as `Int?` if we want to check for its nullity later.

File: syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+         if (numSamples != null) {
+            return numSamples
+         }
+         return 0
+    }
+
+    override fun iter(){}
+
+    override fun len(): Int {
+        return numSamples
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt:31]: What is the difference between `len()` and `numsamples()`?
Shouldn't be `len()` the size of the dataset instead?

File: syftlib/src/main/java/org/openmined/syft/datasource/BatchSampler.kt [hunk 9]
```diff
@@ -0,0 +1,11 @@
+package org.openmined.syft.datasource
+
+/**
+ * Wraps another sampler to yield a mini-batch of indices.
+ * @param sampler (Sampler or Iterable): Base sampler. Can be any iterable object
+ * @param batchSize (Int): Size of mini-batch.
+ * @param dropLast (Boolean): If ``True``, the sampler will drop the last batch if
+ *                          its size would be less than ``batchSize``
+ */
+class BatchSampler (var sampler: Sampler, var batchSize: Int = 1, var dropLast: Boolean = false){
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/BatchSampler.kt:10]: Prefer immutability of params (`val`).

File: syftlib/src/main/java/org/openmined/syft/datasource/SequentialSampler.kt [hunk 10]
```diff
@@ -0,0 +1,14 @@
+package org.openmined.syft.datasource
+
+/**
+ * Samples elements sequentially, always in the same order.
+ * @param dataSource (Dataset): dataset to sample from
+ */
+class SequentialSampler(override var dataSource: Dataset) : Sampler(dataSource) {
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/SequentialSampler.kt:7]: Prefer immutability (use `val`).

**Review by mccorby — CHANGES_REQUESTED** [review]

File: syftlib/src/main/java/org/openmined/syft/datasource/DataLoader.kt (hunk 5 above)
Comment (pengyuan-zhou) [reply syftlib/src/main/java/org/openmined/syft/datasource/DataLoader.kt:24]: > Either pass the `sampler` instance or `shuffle` but not both (I'd prefer the instance).

hmm, but this is supposed to allow user to choose to use RandomSampler or SequentialSampler.
Not sure how to do that without the shuffle boolean

Comment (mccorby) [reply syftlib/src/main/java/org/openmined/syft/datasource/DataLoader.kt:24]: The user needs to pass the sampler in the constructor. He/she would have built it before passing it to this class.

File: syftlib/src/main/java/org/openmined/syft/datasource/DataLoader.kt [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+ * @param dropLast (Boolean, optional): set to ``True`` to drop the last incomplete batch,
+ *                  if the dataset size is not divisible by the batch size. If ``False`` and
+ *                  the size of dataset is not divisible by the batch size, then the last batch
+ *                  will be smaller. (default: ``False``)
+ */
+class DataLoader(var dataset: Dataset,
+                 var batchsize: Int = 1,
+                 var sampler: Sampler,
+                 var dropLast: Boolean = false
+                ) : Dataset, Sampler {
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/DataLoader.kt:26]: Is a `DataLoader` a `Dataset` and a `Sampler` too? I think `DataLoader` collaborates with both (as described in the parameters used to build it) but it's an independent class.

File: syftlib/src/main/java/org/openmined/syft/datasource/DataLoader.kt [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+ * @param batchSize (Int, optional): how many samples per batch to load (default: ``1``).
+ * @param sampler (Sampler or Iterable, optional): defines the strategy to draw
+ *                  samples from the dataset. Can be any ``Iterable`` with ``__len__``
+ *                  implemented. If specified, :attr:`shuffle` must not be specified.
+ * @param dropLast (Boolean, optional): set to ``True`` to drop the last incomplete batch,
+ *                  if the dataset size is not divisible by the batch size. If ``False`` and
+ *                  the size of dataset is not divisible by the batch size, then the last batch
+ *                  will be smaller. (default: ``False``)
+ */
+class DataLoader(var dataset: Dataset,
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/DataLoader.kt:22]: Use `val` for the parameters as they must not changed.

File: syftlib/src/main/java/org/openmined/syft/datasource/Dataset.kt [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+
+    /**
+     * This method is called to return the size of the dataset, needs to be overridden.
+     */
+    fun len() : Int {return  0}
+
+    /**
+     * This method is called to fetch a data sample for a given key.
+     */
+    fun getitem(index: Float) {}
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/Dataset.kt:21]: Shouldn't this function return some type?

File: syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt [hunk 14]
```diff
@@ -0,0 +1,35 @@
...
+import org.pytorch.Tensor
+
+/**
+ * Samples elements randomly. If without replacement, then sample from a shuffled dataset.
+ * If with replacement, then user can specify :attr:`num_samples` to draw.
+ * @param dataSource (Dataset): dataset to sample from
+ * @param numSamples (Int): number of samples to draw, default=`len(dataset)`. This argument
+ *                          is supposed to be specified only when `replacement` is ``True``.
+ */
+class RandomSampler(val dataSource: Dataset, val numSamples: Int) :  Dataset, Sampler {
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/RandomSampler.kt:13]: `RandomSampler` should not be of type `Dataset`.

File: syftlib/src/main/java/org/openmined/syft/datasource/Sampler.kt [hunk 15]
```diff
@@ -0,0 +1,12 @@
+package org.openmined.syft.datasource
+
+/**
+ * Base class for all Samplers.
+ * Every Sampler subclass has to provide an :meth:`__iter__` method, providing a
+ * way to iterate over indices of dataset elements, and a :meth:`__len__` method
+ * that returns the length of the returned iterators.
+ */
+interface Sampler: Dataset {
```
Comment (mccorby) [inline syftlib/src/main/java/org/openmined/syft/datasource/Sampler.kt:9]: Why is `Sampler` a `Dataset`?

File: tutorial/Part 1 - Train on custom dataset.md [hunk 16]
```diff
@@ -0,0 +1,174 @@
+# Part 1 - Train on custom dataset
```
Comment (mccorby) [inline tutorial/Part 1 - Train on custom dataset.md:1]: We can review this file once we have a clear idea of what the objects/classes would be.

**Review by mccorby — CHANGES_REQUESTED** [review]
